aten/src/ATen/core/boxing/KernelFunction.h

```diff
@@ -0,0 +1,466 @@
+   * this if you know the function pointer at compile time, see doc comment
+   * there for an explanation.
+   *
+   * Example:
+   *
+   * > Tensor unboxed_func(Tensor a, Tensor b) {...}
+   * > KernelFunction func = KernelFunction::makeFromUnboxedRuntimeFunction(&unboxed_func);
+   */
+  template<bool AllowLegacyTypes = false, class FuncType>
+  static KernelFunction makeFromUnboxedRuntimeFunction(FuncType* func) {
+    static_assert(guts::is_function_type<FuncType>::value, "Tried to call KernelFunction::makeFromUnboxedRuntimeFunction with a non-function type.");
+    static_assert(!std::is_same<FuncType, BoxedKernelFunction>::value, "Tried to call KernelFunction::makeFromUnboxedRuntimeFunction with a boxed function pointer. Please use KernelFunction::makeFromBoxedFunction instead.");
+    TORCH_INTERNAL_ASSERT(func != nullptr, "Kernel function cannot be nullptr");
+
+    return makeFromUnboxedFunctor<AllowLegacyTypes>(
+      std::make_shared<detail::WrapRuntimeKernelFunctor<guts::decay_t<FuncType>>>(func)
+    );
+  }
+
+  /**
+   * Create a KernelFunction from an unboxed lambda.
+   *
+   * Example:
+   *
+   * > KernelFunction func = KernelFunction::makeFromUnboxedLambda(
+   * >      [] (Tensor a, bool b) -> Tensor {...});
+   */
+  template<bool AllowLegacyTypes = false, class Lambda>
+  static KernelFunction makeFromUnboxedLambda(Lambda&& lambda) {
+    static_assert(guts::is_functor<guts::decay_t<Lambda>>::value, "Tried to call KernelFunction::makeFromUnboxedLambda with a non-lambda type.");
+
+    return makeFromUnboxedFunctor<AllowLegacyTypes>(
+      std::make_shared<detail::WrapRuntimeKernelFunctor<guts::decay_t<Lambda>>>(std::forward<Lambda>(lambda))
+    );
+  }
+
+private:
+
+  explicit KernelFunction(std::function<std::shared_ptr<OperatorKernel>()> functorFactory, std::shared_ptr<OperatorKernel> functor, BoxedKernelFunction* boxed_kernel_func, void* unboxed_kernel_func, c10::optional<uint64_t> signature_hash)
+  : functorFactory_(std::move(functorFactory))
+  , functor_(std::move(functor))
+  , boxed_kernel_func_(boxed_kernel_func)
+  , unboxed_kernel_func_(unboxed_kernel_func)
+  , signature_hash_(signature_hash)
+  {}
+
+  OperatorKernel* getFunctor_() const {
+    if (functor_.get() == nullptr) {
+      if (!functorFactory_) {
+        return nullptr;
+      }
+      functor_ = functorFactory_();
+    }
+    return functor_.get();
+  }
+
+  // If the operator has an unboxed_kernel_func, then either
+  // functorFactory_ or functor_ must be set, possibly both.
+  // If functor_ is not set but functorFactory_ is, we will create
+  // functor_ by calling functorFactory_ the first time it is needed.
+  // We use this indirection because many KernelFunctions are created
+  // at static initialization time but are created with functors that
+  // store Tensor and we can't call the Tensor() constructor at static
+  // initialization time yet (SIOF). So these register with a
```

Comment:
Err, really?

Reply:
(I am specifically confused about the assertion that many functors store tensor)

Reply:
I think you're right. This statement was true in the caffe2 world but is likely not true anymore in the PyTorch world. I changed it to "because we want to allow functors to store Tensors".

Reply:
Even in C2 world it's not true - we wouldn't create an instance of operator at registration time and lifetime of operator would be managed by the invocation context, not centralized dispatcher. I honestly don't think there's a usecase for kernelFactory_ today (and in the earlier design it was not fully figured out either). Can't we just nuke the functorFactory and just declare that it's always functor?

Reply:
So we're saying it's never ok for a kernel functor to store Tensors? I feel uneasy about that since it seems an obvious thing to store. For example, one of the first designs for how to handle constants in the lite interpreter was to register a custom constant operator on the fly for each constant in a model when we load that model. That constant operator kernel would need to store the Tensor value it represents. The lite interpreter design changed and we don't need it anymore, but storing tensors still seems like an obviously useful feature.

If it really turns out to be not used, we can easily change this later. The existence of `kernelFactory_` is well hidden behind the custom op API, users don't see it.

Reply:
I'm advocating for an even more hardline stance: NO closures, except for generated boxed wrappers (which need to know the pointer of the thing they're wrapping).

> Removing it now would require some changes to the experimental caffe2 ops I exported to c10

Can we just delete them? :)

Reply:
Yes we could. Same point as above though, that would be a change in functionality and is unrelated to this PR. Let's set up a separate meeting to discuss this and not block this PR on it.